server/src/HostResourceQueryOption.cc

```diff
@@ -436,19 +548,227 @@ string HostResourceQueryOption::makeConditionForNormalUser(
 		++numServers;
 	}
 
-	if (targetHostId != ALL_LOCAL_HOSTS) {
-		DBTermCStringProvider rhs(*getDBTermCodec());
-		return StringUtils::sprintf(
-		         "((%s) AND %s=%s)",
-		         condition.c_str(), hostIdColumnName.c_str(),
-		         rhs(targetHostId));
-	}
-
 	if (numServers == 1)
 		return condition;
 	return StringUtils::sprintf("(%s)", condition.c_str());
 }
 
+string HostResourceQueryOption::makeConditionSelectedServers(void) const
+{
+	if (m_impl->selectedServerIdSet.empty())
+		return string();
+
+	DBTermCStringProvider rhs(*getDBTermCodec());
+	string condition;
+
+	for (auto &serverId: m_impl->selectedServerIdSet) {
+		if (!isAllowedServer(serverId))
+			continue;
+		addCondition(condition,
+			     StringUtils::sprintf(
+			       "%s=%s",
+			       getServerIdColumnName().c_str(),
+			       rhs(serverId)),
+			     ADD_TYPE_OR);
+	}
+
+	if (m_impl->selectedServerIdSet.size() > 1)
+		return StringUtils::sprintf("(%s)", condition.c_str());
+	else
+		return condition;
+}
+
+string HostResourceQueryOption::makeConditionExcludedServers(void) const
+{
+	if (m_impl->excludedServerIdSet.empty())
+		return string();
+
+	DBTermCStringProvider rhs(*getDBTermCodec());
+	string condition;
+
+	for (auto &serverId: m_impl->excludedServerIdSet) {
+		if (!isAllowedServer(serverId))
+			continue;
+		addCondition(condition,
+			     StringUtils::sprintf(
+			       "%s<>%s",
+			       getServerIdColumnName().c_str(),
+			       rhs(serverId)));
+	}
+
+	if (m_impl->excludedServerIdSet.size() > 1)
+		return StringUtils::sprintf("(%s)", condition.c_str());
+	else
+		return condition;
+}
+
+string HostResourceQueryOption::makeConditionSelectedHostgroups(void) const
+{
+	if (m_impl->selectedServerHostgroupSetMap.empty())
+		return string();
+
+	DBTermCStringProvider rhs(*getDBTermCodec());
+	string condition;
+	string serverIdColumnName = getServerIdColumnName();
+	string hostgroupIdColumnName = getHostgroupIdColumnName();
+
+	for (auto &pair: m_impl->selectedServerHostgroupSetMap) {
+		if (!isAllowedServer(pair.first))
+			continue;
+		for (auto &hostgroupId: pair.second) {
+			if (!isAllowedHostgroup(pair.first, hostgroupId))
+				continue;
+			addCondition(condition,
```

Comment:
This is just an impression. No need to take care of it.

We want to define a new class that handles a condition string, provides utility methods to concatenate addition conditions and returns C style string pointer.